File: src/main/c/groovyclient.c
/*
 * Copyright 2009-2013 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#if defined(_MSC_VER) || defined(__MINGW32__)
#define WINDOWS
#else
#define UNIX
#endif

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <sys/param.h>
#include <unistd.h>
#include <signal.h>

#include "config.h"

#ifdef WINDOWS
#include <windows.h>
#include <winsock2.h>
#include <process.h>
#include <sys/fcntl.h>
#endif

#include "option.h"
#include "session.h"

static void scriptdir(char* result_dir, char* script_path)
{
    // prepare work variable of script path
    int script_path_length = strlen(script_path);
    char work_path[script_path_length];
    strcpy(work_path, script_path);

    // cut off a part of script name
    char* work_pt = work_path + script_path_length;
    while (work_pt > work_path && (*work_pt != '/' && *work_pt != '\\')) {
        work_pt--;
    }
    if (*work_pt == '/' || *work_pt == '\\') {
        work_pt++;
    }
    if (*work_pt != '\0') { // cut
        *work_pt = '\0';
    }

    // set result
    strcpy(result_dir, work_path);

#ifdef DEBUG
    fprintf(stderr, "DEBUG: scriptdir: %s, %zu\n", result_dir, strlen(result_dir));
#endif
}

static char* groovyserver_cmdline(char* script_path, char* arg, int port)
{
    // resolve base directory
    char basedir_path[MAXPATHLEN];
    char* groovyserv_home = getenv("GROOVYSERV_HOME");
    if (groovyserv_home == NULL) {
        scriptdir(basedir_path, script_path);
    } else {
#ifdef WINDOWS
        // If the path is on cygwin, scriptdir is used instead of GROOVYSERV_HOME.
        if (*groovyserv_home == '/') {
            if (!client_option.quiet) {
                fprintf(stderr, "WARN: Using the directory path of groovyclient instead of the invalid GROOVYSERV_HOME which isn't a path format on Windows: %s\n", groovyserv_home);
                fflush(stderr);
            }
            scriptdir(basedir_path, script_path);
        } else {
            sprintf(basedir_path, "%s\\bin\\", groovyserv_home);
        }
#else
        sprintf(basedir_path, "%s/bin/", groovyserv_home);
#endif
    }

#ifdef DEBUG
    fprintf(stderr, "DEBUG: basedir_path: %s, %zu\n", basedir_path, strlen(basedir_path));
#endif

    // make command line to invoke groovyserver
    static char cmdline[MAXPATHLEN];
#ifdef WINDOWS
    sprintf(cmdline, "\"%sgroovyserver.bat\" -p %d %s", basedir_path, port, arg);
#else
    sprintf(cmdline, "'%sgroovyserver' -p %d %s", basedir_path, port, arg);
#endif

#ifdef DEBUG
    fprintf(stderr, "DEBUG: cmdline: %s, %zu\n", cmdline, strlen(cmdline));
#endif

    return cmdline;
}

void invoke_server(char* script_path, int port, char* main_opt, char* authtoken)
{
    char* opt = strdup(main_opt);
    char* quiet_option = " -q";

    if (client_option.quiet) {
        opt = realloc(opt, strlen(opt) + strlen(quiet_option) + 1);
        strcat(opt, quiet_option);
    }

    char authtoken_option[BUFFER_SIZE];
    if (authtoken != NULL) {
        sprintf(authtoken_option, " --authtoken %s", authtoken);
        opt = realloc(opt, strlen(opt) + strlen(authtoken_option) + 1);
        strcat(opt, authtoken_option);
    }

    char* cmdline = groovyserver_cmdline(script_path, opt, port);
    if (!client_option.quiet) {
        fprintf(stderr, "Invoking server: %s\n", cmdline);
        fflush(stderr);
    }
    system(cmdline);
    free(opt);
}

void start_server(char* script_path, int port, char* authtoken)
{
    invoke_server(script_path, port, "", authtoken);
}

void kill_server(char* script_path, int port)
{
    invoke_server(script_path, port, "-k", NULL);
}

void restart_server(char* script_path, int port, char* authtoken)
{
    invoke_server(script_path, port, "-r", authtoken);
}

/*
 * read authentication authtoken.
 */
static void read_authtoken(char* authtoken, int size, int port)
{
    char path[MAXPATHLEN];
#ifdef WINDOWS
    sprintf(path, "%s\\.groovy\\groovyserv\\authtoken-%d", getenv("USERPROFILE"), port);
#else
    sprintf(path, "%s/.groovy/groovyserv/authtoken-%d", getenv("HOME"), port);
#endif
    FILE* fp = fopen(path, "r");
    if (fp != NULL) {
        if (fgets(authtoken, size, fp) == NULL) {
            perror("ERROR: fgets");
            exit(1);
        }
        fclose(fp);
    }
    else {
        fprintf(stderr, "ERROR: cannot open authtoken file\n");
        exit(1);
    }
}

static char* get_host()
{
    if (client_option.host != NULL) {
        return client_option.host;
    }

    static char* host_env;
    host_env = getenv("GROOVYSERVER_HOST");
    if (host_env != NULL) {
        return host_env;
    }

    return DESTHOST;
}

static int get_port()
{
    if (client_option.port != PORT_NOT_SPECIFIED) {
        return client_option.port;
    }

    char* port_str = getenv("GROOVYSERVER_PORT");
    if (port_str != NULL) {
        int port;
        if (sscanf(port_str, "%d", &port) != 1) {
            fprintf(stderr, "ERROR: port number %s of GROOVYSERV_PORT error\n", port_str);
            exit(1);
        }
        return port;
    }

    return DESTPORT;
}

static char* get_authtoken_specified_by_client(int port)
{
    if (client_option.authtoken != NULL) {
        return client_option.authtoken;
    }
    return NULL;
}

static char* get_authtoken_generated_by_server(int port)
{
    static char authtoken[BUFFER_SIZE];
    read_authtoken(authtoken, sizeof(authtoken), port);
    return authtoken;
}

static int connect_server(char* argv0, char* host, int port, char* authtoken)
{
    int fd;

    int failCount = 0;

    while ((fd = open_socket(host, port)) == -1) {
        if (failCount >= 3) {
            fprintf(stderr, "ERROR: failed to start up groovyserver\n");
            exit(1);
        }
        start_server(argv0, port, authtoken);

#ifdef WINDOWS
        Sleep(3000);
#else
        sleep(3);
#endif
        failCount++;
    }
    return fd;
}

static int fd_soc;

static void signal_handler(int sig) {
#ifdef WINDOWS
    send(fd_soc, "Size: -1\n\n", 10, 0);
    closesocket(fd_soc);
#else
    write(fd_soc, "Size: -1\n\n", 10);
    close(fd_soc);
#endif
    exit(1);
}

/*
 * open socket and initiate session.
 */
int main(int argc, char** argv)
{
#ifdef WINDOWS
    WSADATA wsadata;
    if (WSAStartup(MAKEWORD(1,1), &wsadata) == SOCKET_ERROR) {
        fprintf(stderr, "ERROR: creating socket\n");
        exit(1);
    }

    // make standard output to binary mode.
    if (_setmode(_fileno(stdout), _O_BINARY) < 0) {
        fprintf(stderr, "ERROR: setmode stdout failed\n");
        exit(1);
    }
#endif

    scan_options(&client_option, argc, argv);

    char* host = get_host();
    int port = get_port();

    // authtoken sprintf by client can use for both control to server and connection of server.
    // when authtoken is not specified by client, it means that a user want to use remote
    // groovyserver, groovyserver must generate new authtoken by server side
    char* authtoken = get_authtoken_specified_by_client(port);

    // control server
    if (client_option.kill) {
        kill_server(argv[0], port);
        exit(0);
    }
    else if (client_option.restart) {
        restart_server(argv[0], port, authtoken);
    }

    // connect to server
    fd_soc = connect_server(argv[0], host, port, authtoken);
    signal(SIGINT, signal_handler); // using fd_soc in handler

    // ~/.groovy/groovserv/authtoken-NNNN file data can use only for connection to server.
    // but if authtoken is already specified by client, it must be used to the connection
    // instead of the file data, because it means that a user want to use a remote groovyserver.
    if (authtoken == NULL) {
        authtoken = get_authtoken_generated_by_server(port);
    }

    // invoke a script on server
    send_header(fd_soc, argc, argv, authtoken);
    int status = start_session(fd_soc);

    // print particular error status message
    // FIXME it's strongly bound to exit code of ExitStatus on groovyserver.
    //       and it will easily conflict with exit code which is specified at user script...
    if (status == ERROR_INVALID_AUTHTOKEN) {
        fprintf(stderr, "ERROR: rejected by groovyserv because of invalid authtoken\n");
    } else if (status == ERROR_CLIENT_NOT_ALLOWED) {
        fprintf(stderr, "ERROR: rejected by groovyserv because of not allowed client address\n");
    }

#ifdef DEBUG
    fprintf(stderr, "DEBUG: exit code: %d\n", status);
#endif

    // an additional text after invoking
    if (client_option.help) {
        usage();
        exit(status);
    }
    if (client_option.version) {
        version();
        exit(status);
    }

#ifdef WINDOWS
    WSACleanup();
#endif

    return status;
}

File: src/main/go/cmd/groovyclient/groovyclient.go
/*
 * Copyright 2014 the original author or authors.
 *
 * Licensed under the Apache License, version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package main

import (
	cmn "../../common"
	srv "../../server"
	"fmt"
	"log"
	"os"
	"regexp"
	"strconv"
	"strings"
)

const (
	USAGE = `usage: groovyclient -C[option for groovyclient] [args/options for groovy]
options:
  -Ch,-Chelp                       show this usage
  -Cs,-Chost                       specify host to connect to server
  -Cp,-Cport <port>                specify port to connect to server
  -Ca,-Cauthtoken <authtoken>      specify authtoken
  -Ck,-Ckill-server                kill the running server
  -Cr,-Crestart-server             restart the running server
  -Ct,-Ctimeout <second>           specify a timeout waiting for starting up a
                                   server process (default: 20 sec)
  -Cq,-Cquiet                      suppress starting messages
  -Cenv <substr>                   pass environment variables of which a name
                                   includes specified substr
  -Cenv-all                        pass all environment variables
  -Cenv-exclude <substr>           don't pass environment variables of which a
                                   name includes specified substr
  -Cv,-Cversion                    display the version
  -Ckeep-server-cwd                avoid to change directory to current working
                                   directory of client
  -Cdebug                          display console log`
)

var (
	VersionMessage    = "GroovyServ Version: Client: " + GroovyServVersion
	GroovyServVersion = "X.X.X-SNAPSHOT" // replaced by ldflags
)

type (
	ClientOptions struct {
		host           string
		port           int
		authToken      string
		quiet          bool
		envAll         bool
		envIncludeMask []string
		envExcludeMask []string
		help           bool
		version        bool
		debug          bool
		keepServerCwd  bool
	}

	ServerOptions struct {
		help    bool
		version bool
		args    []string
		kill    bool
		restart bool
		timeout int
	}

	Options struct {
		client ClientOptions
		server ServerOptions
	}
)

func NewOptions() *Options {
	return &Options{
		client: ClientOptions{
			host:           cmn.Env("GROOVYSERV_HOST", srv.DefaultHost),
			port:           cmn.EnvInt("GROOVYSERV_PORT", srv.DefaultPort),
			authToken:      "",
			quiet:          false,
			envAll:         false,
			envIncludeMask: []string{},
			envExcludeMask: []string{},
			help:           false,
			version:        false,
			debug:          false,
			keepServerCwd:  false,
		},
		server: ServerOptions{
			help:    false,
			version: false,
			args:    []string{},
			kill:    false,
			restart: false,
			timeout: srv.DefaultTimeout,
		},
	}
}

func ParseOptions(args cmn.Args) *Options {
	opts := NewOptions()
	for i := 1; i < len(args); i++ {
		arg, param := args[i], ""

		// Replacing "-v.*" -> "--version". This behavior is as same as regular groovy command.
		if regexp.MustCompile(`^-v.*`).MatchString(arg) {
			arg = "-v"
		}

		switch arg {
		case "-Cs", "-Chost":
			i, param = args.Param(i, arg)
			opts.client.host = param
		case "-Cp", "-Cport":
			i, param = args.Param(i, arg)
			port, err := strconv.Atoi(param)
			if err != nil {
				panic(fmt.Sprintf("could not parse port number '%s'", param))
			}
			opts.client.port = port
		case "-Ca", "-Cauthtoken":
			i, param = args.Param(i, arg)
			opts.client.authToken = param
		case "-Cq", "-Cquiet":
			opts.client.quiet = true
		case "-Cenv-all":
			opts.client.envAll = true
		case "-Cenv", "-Cenv-exclude":
			i, param = args.Param(i, arg)
			if arg == "-Cenv" {
				opts.client.envIncludeMask = append(opts.client.envIncludeMask, param)
			} else {
				opts.client.envExcludeMask = append(opts.client.envExcludeMask, param)
			}
		case "-Ch", "-Chelp":
			opts.client.help = true
		case "-h", "--help", "-help":
			opts.server.help = true
			opts.server.args = append(opts.server.args, arg)
		case "-Cv", "-Cversion":
			opts.client.version = true
		case "--version", "-v":
			opts.server.version = true
			opts.server.args = append(opts.server.args, arg)
		case "-Cdebug":
			opts.client.debug = true
		case "-Ckeep-server-cwd":
			opts.client.keepServerCwd = true
		case "-Ck", "-Ckill-server":
			opts.server.kill = true
		case "-Cr", "-Crestart-server":
			opts.server.restart = true
		case "-Ct", "-Ctimeout":
			i, param = args.Param(i, arg)
			timeout, err := strconv.Atoi(param)
			if err != nil {
				panic(fmt.Sprintf("could not parse timeout '%s'", param))
			}
			opts.server.timeout = timeout * 1000
		default:
			if m, _ := regexp.MatchString("^-C.*", arg); m {
				panic(fmt.Sprintf("unrecognized option %s", arg))
			}
			opts.server.args = append(opts.server.args, arg)
		}
	}
	if opts.server.kill && opts.server.restart {
		panic(fmt.Sprintf("invalid arguments: %s\nHint: You cannot specify -Ckill-server and -Crestart-server options at the same time.", strings.Join(args[1:], " ")))
	}
	// To show a help usage if no args for server.
	if len(opts.server.args) == 0 {
		opts.server.help = true
	}
	return opts
}

func main() {
	defer cmn.ExitIfPanic()

	// Parsing options
	opts := ParseOptions(os.Args)

	// Setting log level
	cmn.SetupGlobalLogger(!opts.client.debug)
	log.Println("Original arguments:", os.Args)
	log.Printf("Parsed options: %#v", opts) // must be after affecting -Cdebug option

	// Only show usage or version (highest priority)
	switch {
	case opts.client.help:
		fmt.Println(USAGE)
		os.Exit(0)
	case opts.client.version:
		fmt.Println(VersionMessage)
		os.Exit(0)
	}

	// Setting up a server for running a process
	var server = srv.Server{
		Args:           opts.server.args,
		Host:           opts.client.host,
		Port:           opts.client.port,
		AllowFrom:      "",
		AuthToken:      opts.client.authToken,
		Quiet:          opts.client.quiet,
		EnvAll:         opts.client.envAll,
		EnvIncludeMask: opts.client.envIncludeMask,
		EnvExcludeMask: opts.client.envExcludeMask,
		Help:           opts.client.help,
		Version:        opts.client.version,
		Verbose:        false,
		Debug:          opts.client.debug,
		KeepServerCwd:  opts.client.keepServerCwd,
	}

	// Shutting down a server when specified
	if opts.server.kill || opts.server.restart {
		if err := server.Shutdown(); err != nil {
			panic(err.Error())
		}
		if opts.server.kill {
			os.Exit(0)
		}
	}

	// Starting a server if not running
	if server.Dead() {
		if err := server.Start(opts.server.timeout); err != nil {
			panic(err.Error())
		}
	}

	// Make a groovy script evaluate on a server
	statusCode, err := server.RunScript()
	if err != nil {
		panic(err.Error())
	}

	// Only show usage or version (highest priority)
	switch {
	case opts.server.help:
		fmt.Println("")
		fmt.Println(USAGE)
	case opts.server.version:
		fmt.Println(VersionMessage)
	}

	os.Exit(statusCode)
}
